### PR TITLE
upipe-modules: Add audio_merge pipe

### PR DESCRIPTION
**Pull request by wlcx**
Adds audio_merge pipe which accepts planar audio from input subpipes and outputs a single stream of multiplanar audio.
Adds test for audio_merge pipe

### CODE REVIEW

**Comment by kierank** [comment]
Seeing as float audio is between -1 and +1, you should use more realistic (but nonetheless precise) values in that range.

File: lib/upipe-modules/upipe_audio_merge.c [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+ * @param command type of command to process
+ * @param args arguments of the command
+ * @return an error code
+ */
+static int upipe_audio_merge_sub_control(struct upipe *upipe,
+                                         int command, va_list args)
+{
+    UBASE_HANDLED_RETURN(upipe_audio_merge_sub_control_super(upipe, command, args));
+    switch (command) {
+        case UPIPE_REGISTER_REQUEST: {
```
Comment (quarium) [inline lib/upipe-modules/upipe_audio_merge.c]: You can use the helper `UBASE_HANDLED_RETURN(upipe_control_provide_request(upipe, command, args));`

File: lib/upipe-modules/upipe_audio_merge.c [hunk 2]
```diff
@@ -0,0 +1,638 @@
...
+            struct urequest *request = va_arg(args, struct urequest *);
+            return upipe_throw_provide_request(upipe, request);
+        }
+        case UPIPE_UNREGISTER_REQUEST:
+            return UBASE_ERR_NONE;
+        case UPIPE_SUB_GET_SUPER: {
+            struct upipe **p = va_arg(args, struct upipe **);
```
Comment (quarium) [inline lib/upipe-modules/upipe_audio_merge.c]: You can use the helper `UBASE_HANDLED_RETURN(upipe_audio_merge_sub_control_super(upipe, command, args));`

File: lib/upipe-modules/upipe_audio_merge.c [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+    upipe_audio_merge_init_upump_mgr(upipe);
+    upipe_audio_merge_init_upump(upipe);
+    upipe_audio_merge_init_output(upipe);
+    upipe_audio_merge_init_sub_mgr(upipe);
+    upipe_audio_merge_init_sub_inputs(upipe);
+    upipe_audio_merge_init_ubuf_mgr(upipe);
+
+    upipe_audio_merge->sub_flow_def = NULL;
+
+    upipe_audio_merge_store_flow_def(upipe, flow_def);
```
Comment (quarium) [inline lib/upipe-modules/upipe_audio_merge.c]: This helper will throw an event, so you'll have to do that after the `upipe_throw_ready(upipe);`

File: lib/upipe-modules/upipe_audio_merge.c [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+
+    upipe_audio_merge_check_upump_mgr(upipe);
```
Comment (quarium) [inline lib/upipe-modules/upipe_audio_merge.c]: Same here

File: lib/upipe-modules/upipe_audio_merge.c [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+    UBASE_HANDLED_RETURN(upipe_audio_merge_control_inputs(upipe, command, args));
+    UBASE_HANDLED_RETURN(upipe_audio_merge_control_output(upipe, command, args));
+
+    switch (command) {
+        case UPIPE_ATTACH_UPUMP_MGR:
+            upipe_audio_merge_set_upump(upipe, NULL);
+            return upipe_audio_merge_attach_upump_mgr(upipe);
+        case UPIPE_UNREGISTER_REQUEST:
+            return UBASE_ERR_NONE;
+        case UPIPE_GET_OUTPUT: {
```
Comment (quarium) [inline lib/upipe-modules/upipe_audio_merge.c]: This is already handled by `UBASE_HANDLED_RETURN(upipe_audio_merge_control_output(upipe, command, args));`

File: lib/upipe-modules/upipe_audio_merge.c [hunk 6]
```diff
@@ -0,0 +1,638 @@
...
+            struct upipe **p = va_arg(args, struct upipe **);
+            return upipe_audio_merge_get_output(upipe, p);
+        }
+        case UPIPE_SET_OUTPUT: {
```
Comment (quarium) [inline lib/upipe-modules/upipe_audio_merge.c]: Same here

File: lib/upipe-modules/upipe_audio_merge.c [hunk 7]
```diff
@@ -0,0 +1,638 @@
...
+                                     int command, va_list args)
+{
+    UBASE_HANDLED_RETURN(upipe_audio_merge_control_inputs(upipe, command, args));
+    UBASE_HANDLED_RETURN(upipe_audio_merge_control_output(upipe, command, args));
+
+    switch (command) {
+        case UPIPE_ATTACH_UPUMP_MGR:
+            upipe_audio_merge_set_upump(upipe, NULL);
+            return upipe_audio_merge_attach_upump_mgr(upipe);
+        case UPIPE_UNREGISTER_REQUEST:
```
Comment (quarium) [inline lib/upipe-modules/upipe_audio_merge.c]: This is already handled by `UBASE_HANDLED_RETURN(upipe_audio_merge_control_output(upipe, command, args));`. If you want to override the default behavior, you'll need to handle the control commands before calling the helper.

**Comment by wlcx** [comment]
@quarium Thanks for the review, have updated

File: lib/upipe-modules/upipe_audio_merge.c [hunk 8]
```diff
@@ -0,0 +1,618 @@
...
+        uint8_t output_channels = 0;
+        UBASE_ERROR(upipe, uref_sound_flow_get_channels(upipe_audio_merge->flow_def, &output_channels));
+
+        // If the samples of the input uref != our output sample size, throw an error and don't copy
+        if (unlikely(input_num_samples != output_num_samples))
+            upipe_err_va(upipe, "input samples (%"PRIu64") != output samples (%"PRIu64")!",
+                input_num_samples, output_num_samples);
+        // Todo: ensure #planes in uref matches the flowdef?
+        else {
+            for ( i = 0; i < planes; i++ ) {
```
Comment (kierank) [inline lib/upipe-modules/upipe_audio_merge.c]: spacing

File: lib/upipe-modules/upipe_audio_merge.c [hunk 9]
```diff
@@ -0,0 +1,618 @@
...
+    float *out_data[output_channels];
+
+    /* Alloc and zero the output ubuf */
+    ubuf = ubuf_sound_alloc(upipe_audio_merge->ubuf_mgr, output_num_samples);
+    if (unlikely(ubuf == NULL)) {
+        upipe_throw_error(upipe, UBASE_ERR_ALLOC);
+        return;
+    }
+    UBASE_ERROR(upipe, ubuf_sound_write_float(ubuf, 0, -1, out_data, output_channels));
+    for ( i = 0; i < output_channels; i++ )
```
Comment (kierank) [inline lib/upipe-modules/upipe_audio_merge.c]: spacing

File: lib/upipe-modules/upipe_audio_merge.c [hunk 10]
```diff
@@ -0,0 +1,618 @@
...
+        uint8_t planes = 0;
+        UBASE_ERROR(upipe, uref_sound_flow_get_planes(upipe_audio_merge_sub->flow_def, &planes));
+
+        float *in_data[planes];
+        UBASE_ERROR(upipe, uref_sound_read_float(upipe_audio_merge_sub->uref, 0, -1, (const float**)in_data, planes));
+
+        uint64_t input_num_samples = 0;
+        UBASE_ERROR(upipe, uref_sound_flow_get_samples(upipe_audio_merge_sub->uref, &input_num_samples));
+
+        uint64_t output_num_samples = 0;
```
Comment (kierank) [inline lib/upipe-modules/upipe_audio_merge.c]: Anything flow_def related can be obtained outside the loop.

File: lib/upipe-modules/upipe_audio_merge.c [hunk 11]
```diff
@@ -0,0 +1,618 @@
...
+
+/** @internal @Checks if two flow defs match, except for channel count
+ *
+ * @param one first uref
+ * @param two second uref
+ * @return true if flow defs match, false if not
+ */
+static bool upipe_audio_merge_match_flowdefs(struct uref *one, struct uref *two)
+{
+    return uref_sound_flow_cmp_rate(one, two) == 0
```
Comment (kierank) [inline lib/upipe-modules/upipe_audio_merge.c]: ubase_check for all of these

Comment (quarium) [reply lib/upipe-modules/upipe_audio_merge.c]: Not for the _cmp_ functions `return 0 if both attributes are absent or identical`

File: lib/upipe-modules/upipe_audio_merge.c [hunk 12]
```diff
@@ -0,0 +1,618 @@
...
+                /* Only copy up to the number of channels in the output flowdef,
+                and thus what we've allocated */
+                if ((out_channel_count + i) < output_channels) {
+                    for ( j = 0; j < input_num_samples; j++) {
+                        out_data[out_channel_count+i][j] = in_data[i][j];
+                    }
+                }
+            }
+        }
+        out_channel_count += planes;
```
Comment (kierank) [inline lib/upipe-modules/upipe_audio_merge.c]: bad variable name, something like cur_plane

File: lib/upipe-modules/upipe_audio_merge.c [hunk 13]
```diff
@@ -0,0 +1,618 @@
...
+                input_num_samples, output_num_samples);
+        // Todo: ensure #planes in uref matches the flowdef?
+        else {
+            for ( i = 0; i < planes; i++ ) {
+                /* Only copy up to the number of channels in the output flowdef,
+                and thus what we've allocated */
+                if ((out_channel_count + i) < output_channels) {
+                    for ( j = 0; j < input_num_samples; j++) {
+                        out_data[out_channel_count+i][j] = in_data[i][j];
+                    }
```
Comment (kierank) [inline lib/upipe-modules/upipe_audio_merge.c]: remove braces

Comment (wlcx) [reply lib/upipe-modules/upipe_audio_merge.c]: removed all the way up to the `else`, lots of nesting but I think that's the most consistent?

File: lib/upipe-modules/upipe_audio_merge.c (hunk 10 above)
Comment (wlcx) [reply lib/upipe-modules/upipe_audio_merge.c]: moved `output_samples` and `output_num_channels` out of the loop

File: lib/upipe-modules/upipe_audio_merge.c [hunk 14]
```diff
@@ -0,0 +1,605 @@
...
+
+        uint8_t planes = 0;
+        UBASE_ERROR(upipe, uref_sound_flow_get_planes(upipe_audio_merge_sub->flow_def, &planes));
+
+        float *in_data[planes];
+        UBASE_ERROR(upipe, uref_sound_read_float(upipe_audio_merge_sub->uref, 0, -1, (const float**)in_data, planes));
+
+        uint64_t input_num_samples = 0;
+        UBASE_ERROR(upipe, uref_sound_flow_get_samples(upipe_audio_merge_sub->uref, &input_num_samples));
+
```
Comment (kierank) [inline lib/upipe-modules/upipe_audio_merge.c]: space

File: lib/upipe-modules/upipe_audio_merge.c [hunk 15]
```diff
@@ -0,0 +1,605 @@
...
+        /* If the samples of the input uref != our output sample size, throw an error and don't copy */
+        if (unlikely(input_num_samples != output_num_samples))
+            upipe_err_va(upipe, "input samples (%"PRIu64") != output samples (%"PRIu64")!",
+                input_num_samples, output_num_samples);
+        else
+            for (int i = 0; i < planes; i++)
+                /* Only copy up to the number of channels in the output flowdef,
+                and thus what we've allocated */
+                if ((cur_plane + i) < output_channels)
```
Comment (kierank) [inline lib/upipe-modules/upipe_audio_merge.c]: braces

File: lib/upipe-modules/upipe_audio_merge.c [hunk 16]
```diff
@@ -0,0 +1,605 @@
...
+    float *out_data[output_channels];
+
+    /* Alloc and zero the output ubuf */
+    ubuf = ubuf_sound_alloc(upipe_audio_merge->ubuf_mgr, output_num_samples);
+    if (unlikely(ubuf == NULL)) {
+        upipe_throw_error(upipe, UBASE_ERR_ALLOC);
+        return;
+    }
+    UBASE_ERROR(upipe, ubuf_sound_write_float(ubuf, 0, -1, out_data, output_channels));
+    for (i = 0; i < output_channels; i++)
```
Comment (kierank) [inline lib/upipe-modules/upipe_audio_merge.c]: int i

File: lib/upipe-modules/upipe_audio_merge.c [hunk 17]
```diff
@@ -0,0 +1,606 @@
...
+    struct uref *flow_def;
+    struct upipe *upipe = upipe_audio_merge_alloc_flow(mgr,
+                            uprobe, signature, args, &flow_def);
+    if (unlikely(upipe == NULL))
+        return NULL;
+
+    struct upipe_audio_merge *upipe_audio_merge =
+                              upipe_audio_merge_from_upipe(upipe);
+    upipe_audio_merge_init_urefcount(upipe);
+    urefcount_init(upipe_audio_merge_to_urefcount_real(upipe_audio_merge),
```
[the remark this answers is not in the document]
Comment (wlcx) [reply lib/upipe-modules/upipe_audio_merge.c]: Done - I think I've used the macro properly. Could you confirm?

**Comment by cmassiot** [comment]
Can you squash you commits?

**Comment by JDarnley** [comment]
One of us will have to take this over and squash the commits.  Can you leave it open for now?

**Comment by cmassiot** [comment]
Sure.

**Pull request closed without merging**